tarpc/src/client.rs

```diff
@@ -344,7 +344,26 @@ where
         self.transport_pin_mut()
             .poll_next(cx)
             .map_err(|e| {
+                self.pending_requests_mut().close();
                 let e = Arc::new(e);
+                // drain sender messages after closing the channel
+                loop {
+                    match self.pending_requests_mut().poll_recv(cx) {
+                        // channel is open or a spurious failure
+                        Poll::Pending => continue,
```

Comment:
Oh, acually, the channel returns Pending when it's empty, I think. https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.poll_recv

So this loop can break on `Poll::Pending | Poll::Ready(None)`.

Reply:
Oh, I see now why you did it this way — basically, the receiver is never supposed to return `Poll::Pending` after closure. Interesting — I guess this is good then!

Reply:
Actually, I think there's some subtlety here: reading the docs for [close](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.close), it looks like outstanding Permits are still allowed to send into the channel. But there's no guarantee when the send will happen, so this loop is actually a blocking call. Since this code path is nonblocking, I think that's a problem we have to solve.

What about spawning a task that drains the channel?

On a separate note, I think there are other errors that could occur besides the error in `pump_read`, and I hadn't made any attempt to drain the requests for those paths. And likewise, errors in other areas (like in pump_write) will not result in the receiver being drained. What if all of this draining was moved into a drop impl for RequestDispatch?